### PR TITLE
008 edit list item

### PR DESCRIPTION
**Pull request by felix-horn**
Please review:
- list items are now editable
- delete button is only shown for selected list item
- enter creates new list item / closes the mobile keyboard
- tests are updated (jest & cypress)

Thank you!

### CODE REVIEW

**Review by leonp5 — COMMENTED** [review]
Except the tests, i have no idea from tests ;), it looks good to me :+1:

File: src/01.UI-Elements/AddButton.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import styled from 'styled-components/macro'
+import AddIcon from '@material-ui/icons/Add'
+
+AddButton.propTypes = {
+  onClick: PropTypes.func.isRequired,
+}
+
+export default function AddButton({ onClick }) {
+  return (
+    <AddButtonStyled onClick={onClick} data-testid="add-button">
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/AddButton.js]: You could think of rearranging the name in order to not get confused by the verb at the beginning. ``ButtonAdd`` maybe and I do not think that you need the styled part. :)

File: src/01.UI-Elements/AddButton.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  background: none;
+  display: flex;
+  align-items: center;
+`
+const AddIconStyled = styled(AddIcon)`
+  margin-right: 20px;
+  color: var(--mid-gray);
+`
+
+const ButtenTextStyled = styled.span`
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/AddButton.js]: Do you mean ``Button``in stead of ``Butten``?

File: src/01.UI-Elements/ListItem.js [hunk 3]
```diff
@@ -1,29 +1,59 @@
...
+        data-testid="title-list-item"
+      />
+      {isDeleteIconShown && (
+        <DeleteButtonStyled onClick={onDelete} data-testid="delete-list-item" />
+      )}
     </ListItemStyled>
   )
+
+  function handleEnter(event) {
+    event.target.blur()
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.js]: Uhh, cool :)

File: src/01.UI-Elements/ListItem.js [hunk 4]
```diff
@@ -1,29 +1,59 @@
...
+    onEnter()
+  }
 }
 
 const ListItemStyled = styled.label`
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.js:59]: I think you could leave all the "styled" parts behind  :)

File: src/01.UI-Elements/ListItem.test.js [hunk 5]
```diff
@@ -4,77 +4,69 @@ import user from '@testing-library/user-event'
...
-    const onDeleteMock = jest.fn()
-    const { getByLabelText } = render(
-      <ListItem title="Milk" onToggle={onToggleMock} onDelete={onDeleteMock} />
+    const onToggleCheckboxMock = jest.fn()
+    const { getByRole } = render(
+      <ListItem onToggleCheckbox={onToggleCheckboxMock} />
     )
-    const checkbox = getByLabelText('Milk')
+    const checkbox = getByRole('checkbox')
+    // const checkbox = getByTestId('checkbox') not working
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.test.js:36]: Should comments remain there?

File: src/01.UI-Elements/ListItem.test.js [hunk 6]
```diff
@@ -4,77 +4,69 @@ import user from '@testing-library/user-event'
...
-        isChecked
-        onDelete={onDeleteMock}
-      />
-    )
-
+    const { getByTestId } = render(<ListItem onDelete={onDeleteMock} />)
     const button = getByTestId('delete-list-item')
     user.click(button)
-
     expect(onDeleteMock).toHaveBeenCalledTimes(1)
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.test.js:48]: Sometimes you write ``.toHaveBeenCalled()`` and sometimes ``.toHaveBeenCalledTimes(1)`` Is there a difference in the case of just one call?

File: src/01.UI-Elements/ListItem.test.js [hunk 7]
```diff
@@ -4,77 +4,69 @@ import user from '@testing-library/user-event'
...
   })
+
+  xit('calls handleChange', () => {
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.test.js:51]: Do you not want to execute this test? There is an "x" at the beginning of the line?

File: src/01.UI-Elements/ListItem.test.js [hunk 8]
```diff
@@ -4,77 +4,69 @@ import user from '@testing-library/user-event'
...
+    const onInputChange = jest.fn()
+    const { getByTestId } = render(<ListItem handleChange={onInputChange} />)
+    const inputField = getByTestId('title-list-item')
+    console.log({ inputField })
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.test.js:55]: Necessary?

File: src/01.UI-Elements/ListItem.test.js [hunk 9]
```diff
@@ -4,77 +4,69 @@ import user from '@testing-library/user-event'
...
+    user.type(inputField, 't')
+    expect(onInputChange).toHaveBeenCalledTimes(1)
+    expect(onInputChange).toHaveBeenCalledWith('t')
+  })
+
+  xit('calls setIsDeleteIconShown', () => {
```
Comment (JanaHaeusler) [inline src/01.UI-Elements/ListItem.test.js:61]: Same "x" here

File: src/02.Components/List.js [hunk 10]
```diff
@@ -6,16 +6,21 @@ import ListItem from '../01.UI-Elements/ListItem'
...
+  deleteListItem: PropTypes.func.isRequired,
+  addListItemOnEnter: PropTypes.func.isRequired,
   checked: PropTypes.bool,
 }
 
 export default function List({
-  listAllocation,
-  checked,
   list,
+  listAllocation,
```
Comment (JanaHaeusler) [inline src/02.Components/List.js:18]: What does this do? :)

File: src/App.js [hunk 11]
```diff
@@ -49,7 +54,7 @@ const UndoButtonStyled = styled(UndoButton)`
   &.shown {
   }
 
-  &.fadeToHide {
+  &.fade {
     animation: 1s fadeOut ease forwards;
```
Comment (JanaHaeusler) [inline src/App.js:57]: Love the cool animated undo button! :)

File: src/hooks/useList.js [hunk 12]
```diff
@@ -29,7 +30,23 @@ export default function useList() {
     visibilityUndoButton,
   }
 
-  function addListItem(title, isChecked = false) {
+  function handleInputChange(event) {
```
Comment (JanaHaeusler) [inline src/hooks/useList.js]: In some functions you just write ``e``and sometimes ``event``

**Review by JanaHaeusler — COMMENTED** [review]
Love all your hard coding work, Felix! :) Impressive list items and thoroughly testing. I left some thoughts and comments. If something is not understandable just let me know. Hope it helps :)

File: .gitignore [hunk 13]
```diff
@@ -2,3 +2,10 @@ node_modules
 dist
 .DS_Store
 build
+cypress/screenshots
+cypress/support
+cypress/fixtures
+cypress/videos
+cypress/plugins
+
```
Comment (monroho) [inline .gitignore:10]: Oopsie :D

File: src/01.UI-Elements/AddButton.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+AddButton.propTypes = {
+  onClick: PropTypes.func.isRequired,
+}
+
+export default function AddButton({ onClick }) {
+  return (
+    <AddButtonStyled onClick={onClick} data-testid="add-button">
+      <AddIconStyled />
+      <ButtenTextStyled>Listeneintrag</ButtenTextStyled>
```
Comment (monroho) [inline src/01.UI-Elements/AddButton.js]: Butten -> Button i guess.

File: src/App.js [hunk 15]
```diff
@@ -1,16 +1,18 @@
 import useList from './hooks/useList'
+import styled from 'styled-components/macro'
 
 import List from './02.Components/List'
-import InputForm from './01.UI-Elements/InputForm'
 import UndoButton from './01.UI-Elements/UndoButton'
-import styled from 'styled-components/macro'
+import AddButton from './01.UI-Elements/AddButton'
```
Comment (monroho) [inline src/App.js]: Why that empty lines between the Imports?

File: src/01.UI-Elements/AddButton.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+  align-items: center;
+`
+const AddIconStyled = styled(AddIcon)`
+  margin-right: 20px;
+  color: var(--mid-gray);
+`
+
+const ButtenTextStyled = styled.span`
+  color: var(--light-gray);
+`
```
Comment (monroho) [inline src/01.UI-Elements/AddButton.js]: You really wanted that whole thing to be a clickable Button? Including the text "Listeneintrag" ? I expected "Listeneintrag" to be smth. like a Placeholder where i can input my "item text", but instead it instantly created an empty Item.

File: src/hooks/useList.js [hunk 17]
```diff
@@ -48,9 +65,9 @@ export default function useList() {
       : setUncheckedIds([...uncheckedIds, newId])
   }
 
-  function undoDelete() {
-    addListItem(deletedListItem.title, deletedListItem.isChecked)
-    setVisibilityUndoButton('hidden')
+  function addListItemOnEnter(targetId) {
```
Comment (monroho) [inline src/hooks/useList.js:68]: Why is there extra logic needed to handle DataInput via Enter Key?

File: src/hooks/useList.js [hunk 18]
```diff
@@ -29,7 +30,23 @@ export default function useList() {
...
+        ...list.byId,
+        [targetId]: {
+          ...list.byId[targetId],
+          title: fieldValue,
+        },
+      },
+    })
+  }
+
+  function addListItem(title = '', isChecked = false) {
```
Comment (monroho) [inline src/hooks/useList.js:49]: You could think about basic validation f.e. not adding a new Item if it`s emtpy?
Right now i just start up the App, click on the Button once and then press enter 5 times and i have 6 empty Items in the List. There is no point in allowing that.

**Review by monroho — COMMENTED** [review]
Saw some really cool ideas here, good job. But the whole logic of you application just lives in that one "useList" hook and you keep adding stuff to it. For me it was really hard to follow the data-flow in your App. Maybe you could think about breaking it down a little, to reduce complexity and make it easier to maintain and test. I also tested your application and def. found some bugs, maybe we can have a look together tomorrow?

File: src/01.UI-Elements/ListItem.test.js (hunk 5 above)
Comment (felix-horn) [reply src/01.UI-Elements/ListItem.test.js:36]: Two of the tests are not working yet. That's why I kept the Xs and comments in there. I'll try to figure it out next week :)

File: src/hooks/useList.js (hunk 12 above)
Comment (felix-horn) [reply src/hooks/useList.js]: Changed it. Thanks.
